### PR TITLE
Implemented safety mechanisms around vault hooks

### PR DESCRIPTION
**Pull request by asselstine**
- Hooks have gas limits
- Hook failures are recorded during batch calls, so they can be avoided.

### CODE REVIEW

File: src/Vault.sol [hunk 1]
```diff
@@ -98,6 +98,12 @@ error LPZeroAddress();
 /// @param maxYieldFeePercentage The max yield fee percentage in integer format (this value is equal to 1 in decimal format)
 error YieldFeePercentageGTPrecision(uint256 yieldFeePercentage, uint256 maxYieldFeePercentage);
 
+error BeforeClaimPrizeFailed(bytes reason);
+
+error AfterClaimPrizeFailed(bytes reason);
+
+uint256 constant HOOK_GAS = 150_000;
```
Comment (PierrickGT) [inline src/Vault.sol]: Add Natspec docs.

File: src/Vault.sol [hunk 2]
```diff
@@ -199,6 +205,13 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
    */
   event RecordedExchangeRate(uint256 exchangeRate);
 
+  event ClaimFailed(
```
Comment (PierrickGT) [inline src/Vault.sol:221]: Add Natspec doc.

File: src/Vault.sol [hunk 3]
```diff
@@ -1038,6 +1058,15 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
   }
 
   /* ============ Claim Functions ============ */
+  function claimPrize(
```
Comment (PierrickGT) [inline src/Vault.sol:1083]: Add Natspec doc.

File: src/Vault.sol [hunk 4]
```diff
@@ -1048,18 +1077,24 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
...
    */
-  function _claimPrize(
+  function claimPrize_INTERNAL_USE_ONLY(
     address _winner,
     uint8 _tier,
     uint32 _prizeIndex,
     uint96 _fee,
     address _feeRecipient
-  ) internal returns (uint256) {
+  ) external returns (uint256) {
```
Comment (PierrickGT) [inline src/Vault.sol:1108]: Should be `internal` no? Or even `private`.

File: src/Vault.sol [hunk 5]
```diff
@@ -1239,4 +1271,9 @@ contract Vault is ERC4626, ERC20Permit, ILiquidationSource, Ownable {
   function _setYieldFeeRecipient(address yieldFeeRecipient_) internal {
     _yieldFeeRecipient = yieldFeeRecipient_;
   }
+
+  modifier onlyClaimer() {
```
Comment (PierrickGT) [inline src/Vault.sol:1300]: Add Natspec doc.

**Review by PierrickGT — CHANGES_REQUESTED** [review]

File: src/Vault.sol (hunk 4 above)
Comment (asselstine) [reply src/Vault.sol:1108]: It has to be external because of the `try catch`.  Kind of a weird trick

**Pull request merged**